cirq/testing/order_tester.py

```diff
@@ -0,0 +1,230 @@
+            (True, NotImplemented),
+            (NotImplemented, True),
+            (NotImplemented, NotImplemented),
+                ]
+
+    def verify_ascending_group(self, *group_items: Any):
+        """Verifies that the given items are strictly ascending
+        with regard to the groups which have been added before.
+        """
+        assert group_items
+
+        for lesser_group in self.groups:
+            for lesser_item in lesser_group:
+                for larger_item in group_items:
+                    self._verify_ascending(lesser_item, larger_item)
+
+
+    def add_ascending(self, *items: Any):
+        """Tries to add a sequence of ascending items to the order tester.
+
+        This methods asserts that items must all be ascending
+        with regard to both each other and the elements which have been already
+        added during previous calls.
+        Some of the previously added elements might be equivalence groups,
+        which are supposed to be equal to each other within that group.
+
+        Args:
+          *items: The sequence of strictly ascending items.
+
+        Raises:
+            AssertionError: Items are not ascending either
+                with regard to each other, or with regard to the elements
+                which have been added before.
+        """
+
+        assert items
+
+        for item in items:
+            self.add_ascending_equivalence_group(item)
+
+    def add_ascending_equivalence_group(self, *group_items: Any):
+        """Tries to add an ascending equivalence group to the order tester.
+
+        Asserts that the group items are equal to each other, but strictly
+        ascending with regard to the already added groups.
+
+        Adds the objects as a group.
+
+        Args:
+            group_items: items making the equivalence group
+
+        Raises:
+            AssertionError: The group elements aren't equal to each other,
+                or items in another group overlap with the new group.
+        """
+        # Check that elements are equal with regard to each other
+        # and not equal to any other group.
+        super(OrderTester, self).verify_equality_group(*group_items)
+
+        # Check that the new group is strictly ascending with regard to
+        # the groups which have been added before.
+        self.verify_ascending_group(*group_items)
+
+        # Remember this group.
+        self.groups.append(tuple(group_items))
+
+
+class ClassSmallerThanEverythingElse:
+    """Assume that the element of this class is less than anything else."""
+
+    def __eq__(self, other):
+        return isinstance(other, ClassSmallerThanEverythingElse)
+
+    def __ne__(self, other):
+        return not isinstance(other, ClassSmallerThanEverythingElse)
+
+    def __lt__(self, other):
+        return not isinstance(other, ClassSmallerThanEverythingElse)
+
+    def __hash__(self):
+        return hash(ClassSmallerThanEverythingElse)  # coverage: ignore
```

Comment:
It shouldn't be necessary to say the coverage of this method is ignored. This method should be indirectly invoked via comparisons within EqualsTester.

Reply:
It didn't do so for the tests I had. It might be because in the existing code in EqualsTester, it only tests hashes after verifying that the new elements aren't equal to anything else which has been added before, which would always fail for that class.